src/dragonball/src/vm/aarch64.rs

```diff
@@ -96,6 +97,30 @@ impl Vm {
         Ok(())
     }
 
+    /// Set boot cpu count info into cmdline
+    #[cfg(feature = "hotplug")]
+    pub fn set_cmdline_cpus(&mut self) -> std::result::Result<(), StartMicroVmError> {
+        let cmdline = self
+            .kernel_config
+            .as_mut()
+            .ok_or(StartMicroVmError::MissingKernelConfig)?
+            .kernel_cmdline_mut();
+
+        cmdline
+            .insert(
+                "dragonball_bootcpus",
+                &format!("{}", self.vm_config.vcpu_count),
+            )
+            .map_err(StartMicroVmError::AddCmdlineCpuCountError)?;
+        cmdline
+            .insert("maxcpus", &format!("{}", self.vm_config.max_vcpu_count))
```

Comment:
IMO, maxcpus means present cpu number not the possible cpu but you want to pass the possible cpu number here. Am I right?

Reply:
"maxcpus" here is used to initialize `setup_max_cpus` variable in kernel. See [maxcpus cmdline](https://github.com/torvalds/linux/blob/f1fcbaa18b28dec10281551dfe6ed3a3ed80e3d6/kernel/smp.c#L920).
According to the comment in kernel, I think you're right. See [setup_max_cpus definition](https://github.com/torvalds/linux/blob/f1fcbaa18b28dec10281551dfe6ed3a3ed80e3d6/kernel/smp.c#L879).

Reply:
Thanks @xuejun-xj -,

So this solusion replies on the "bootvcpu" and "maxvcpu" passed from cmdline which is customized for kata/DB and is hard for a common scenario eg. for a standalone VM.

Is it possible to reuse DT to pass those info then we can extend this feature for a wider use case?

Reply:
Yeah, that's a good idea. Thanks @jongwu :)
As DT contains `max_vcpu_count` cpu nodes, how about adding an attribute to indicate whether this cpu should be activated at booting procedure?
The core problem is to pass the correct "boot_vcpu_count" to bringup_nonboot_cpus function. Do you have any better suggestions about this?
By the way, do you know what's the difference between possible cpu and present cpu? In which case the number of them are unequal.

Reply:
> Yeah, that's a good idea. Thanks @jongwu :) As DT contains `max_vcpu_count` cpu nodes, how about adding an attribute to indicate whether this cpu should be activated at booting procedure?
> 
I think it make sense.
> 
> The core problem is to pass the correct "boot_vcpu_count" to bringup_nonboot_cpus function. Do you have any better suggestions about this?
> 
Maybe this issue can be resolved gracefully by getting boot_cpu number from DT if we can extend use of DT (may violate DT spec)
> 
> By the way, do you know what's the difference between possible cpu and present cpu? And in which case the number of them are unequal?
> 
The meaning for possible & present cpu in x86 is easy to understand. The present cpu means the plugged cpu. The possible cpu means the plugged cpu number plus the cpu number that will be plugged in the future. There is a option "smp" in qemu, like: `-smp 3,sockets=1,cores=15,threads=1,maxcpus=15` where "3" means present cpu number, maxcpus means possible cpu number. At first, there is only 3 vcpu in the guest, for now the preset vcpu is not equal to possible cpu, unless you hotplug 12 more vcpu into guest.
 
As Jame's patch about "support vcpu hotplug on arm64" for linux kernel is under review, where the meaning of possible cpu and present cpu is changed but is confusing on arm side. We can leave it aside for now, I think.

Reply:
@jongwu Thanks for explanation! :)
> Maybe this issue can be resolved gracefully by getting boot_cpu number from DT if we can extend use of DT (may violate DT spec)

I will update the implementation about this extension recently.

> The meaning for possible & present cpu in x86 is easy to understand. The present cpu means the plugged cpu. The possible cpu means the plugged cpu number plus the cpu number that will be plugged in the future. There is a option "smp" in qemu, like: -smp 3,sockets=1,cores=15,threads=1,maxcpus=15 where "3" means present cpu number, maxcpus means possible cpu number. At first, there is only 3 vcpu in the guest, for now the preset vcpu is not equal to possible cpu, unless you hotplug 12 more vcpu into guest.
> 
> As Jame's patch about "support vcpu hotplug on arm64" for linux kernel is under review, where the meaning of possible cpu and present cpu is changed but is confusing on arm side. We can leave it aside for now, I think.

May I understand like this?
- possible_cpu: The max cpu number that can be plugged. Maybe it is decided physical cpu devices?
- present_cpu: The max cpu number that current system supports.
- online_cpu: The cpu number that is currently running on the system.
Am I right?

Reply:
Hi @jongwu. I have thinked of two available solutions:
1. Use a single attribute, like "boot-activated-cpu-number", under /cpus/ node.
2. Use an attribute, like "boot-activated", under each /cpus/cpuX/ node.

IMO, maybe the first solution is better. Is it necessary to use attribute under each /cpus/cpuX? What do you think about it?

Reply:
> @jongwu Thanks for explanation! :)
> 
> > Maybe this issue can be resolved gracefully by getting boot_cpu number from DT if we can extend use of DT (may violate DT spec)
> 
> I will update the implementation about this extension recently.
> 
> > The meaning for possible & present cpu in x86 is easy to understand. The present cpu means the plugged cpu. The possible cpu means the plugged cpu number plus the cpu number that will be plugged in the future. There is a option "smp" in qemu, like: -smp 3,sockets=1,cores=15,threads=1,maxcpus=15 where "3" means present cpu number, maxcpus means possible cpu number. At first, there is only 3 vcpu in the guest, for now the preset vcpu is not equal to possible cpu, unless you hotplug 12 more vcpu into guest.
> > As Jame's patch about "support vcpu hotplug on arm64" for linux kernel is under review, where the meaning of possible cpu and present cpu is changed but is confusing on arm side. We can leave it aside for now, I think.
> 
> May I understand like this?
> 
> * possible_cpu: The max cpu number that can be plugged. Maybe it is decided physical cpu devices?
> * present_cpu: The max cpu number that current system supports.
> * online_cpu: The cpu number that is currently running on the system.
> 
> Am I right?

Here are some reference for you:

- The previous implementation in qemu from Huawei
   - https://patchwork.kernel.org/project/qemu-devel/cover/20200613213629.21984-1-salil.mehta@huawei.com/
- patch in review, kernel from Arm:
   - https://lwn.net/Articles/922127/
- qemu from Huawei
   - https://github.com/salil-mehta/qemu/tree/virt-cpuhp-armv8/rfc-v1-port29092022

Reply:
Another concern is if the implemention here will conflict with the future implemention from kernel. We need take care about it.

Reply:
> Another concern is if the implemention here will conflict with the future implemention from kernel. We need take care about it.

Thanks for your suggestions! I will browse the reference and handle it carefullly.